File: src/services/speechService.ts
export class SpeechService {
  private recognition: SpeechRecognition | null = null;
  private isListening = false;
  private lastRecognizedWords = '';
  private onResultCallback: ((text: string) => void) | null = null;
  private onListeningCallback: (() => void) | null = null;
  private onStoppedCallback: (() => void) | null = null;

  get listening(): boolean {
    return this.isListening;
  }

  get lastResult(): string {
    return this.lastRecognizedWords;
  }

  initialize(): void {
    const SpeechRecognition = (window as unknown as { 
      SpeechRecognition: typeof window.SpeechRecognition; 
      webkitSpeechRecognition: typeof window.SpeechRecognition 
    }).SpeechRecognition ||
      (window as unknown as { webkitSpeechRecognition: typeof window.SpeechRecognition }).webkitSpeechRecognition;

    if (SpeechRecognition) {
      this.recognition = new SpeechRecognition();
      this.recognition.continuous = false;
      this.recognition.interimResults = true;
      this.recognition.lang = 'zh-CN';

      this.recognition.onresult = (event: SpeechRecognitionEvent) => {
        const results = Array.from(event.results);
        const transcript = results
          .map(result => result[0].transcript)
          .join('');
        this.lastRecognizedWords = transcript;
        if (this.onResultCallback) {
          this.onResultCallback(transcript);
        }
      };

      this.recognition.onerror = (event: SpeechRecognitionError) => {
        console.error('Speech recognition error:', event.error);
        this.stopListening();
      };

      this.recognition.onend = () => {
        if (this.isListening) {
          this.isListening = false;
          if (this.onStoppedCallback) {
            this.onStoppedCallback();
          }
        }
      };
    } else {
      console.error('Speech recognition is not supported in this browser');
    }
  }

  startListening(
    onResult: (text: string) => void,
    onListening: () => void,
    onStopped: () => void
  ): void {
    if (!this.recognition) {
      console.error('Speech recognition not initialized');
      return;
    }

    if (this.isListening) {
      this.stopListening();
    }

    this.onResultCallback = onResult;
    this.onListeningCallback = onListening;
    this.onStoppedCallback = onStopped;
    this.isListening = true;

    if (this.onListeningCallback) {
      this.onListeningCallback();
    }

    try {
      this.recognition.start();
    } catch (error) {
      console.error('Failed to start listening:', error);
      this.isListening = false;
    }
  }

  stopListening(): void {
    if (this.recognition && this.isListening) {
      try {
        this.recognition.stop();
      } catch (error) {
        console.error('Failed to stop listening:', error);
      }
      this.isListening = false;
      if (this.onStoppedCallback) {
        this.onStoppedCallback();
      }
    }
  }

  cancelListening(): void {
    if (this.recognition) {
      try {
        this.recognition.abort();
      } catch (error) {
        console.error('Failed to cancel listening:', error);
      }
      this.isListening = false;
      if (this.onStoppedCallback) {
        this.onStoppedCallback();
      }
    }
  }

  speak(text: string): void {
    if ('speechSynthesis' in window) {
      const utterance = new SpeechSynthesisUtterance(text);
      utterance.lang = 'zh-CN';
      utterance.rate = 0.8;
      utterance.volume = 1.0;
      window.speechSynthesis.speak(utterance);
    } else {
      console.error('Text-to-speech is not supported in this browser');
    }
  }

  stopSpeaking(): void {
    if ('speechSynthesis' in window) {
      window.speechSynthesis.cancel();
    }
  }
}

File: lib/services/speech_service.dart
import 'package:speech_to_text/speech_recognition_result.dart';
import 'package:speech_to_text/speech_to_text.dart';
import 'package:flutter_tts/flutter_tts.dart';

class SpeechService {
  final SpeechToText _speechToText = SpeechToText();
  final FlutterTts _flutterTts = FlutterTts();
  
  bool _isListening = false;
  String _lastRecognizedWords = '';
  void Function()? _onStopped;

  bool get isListening => _isListening;
  String get lastRecognizedWords => _lastRecognizedWords;

  Future<void> initialize() async {
    await _speechToText.initialize();
    await _flutterTts.setLanguage('zh-CN');
    await _flutterTts.setSpeechRate(0.8);
    await _flutterTts.setVolume(1.0);
  }

  Future<void> startListening({
    required void Function(String text) onResult,
    required void Function() onListening,
    required void Function() onStopped,
  }) async {
    if (!_speechToText.isAvailable) {
      await _speechToText.initialize();
    }

    if (_speechToText.isListening) {
      await stopListening();
    }

    _isListening = true;
    _onStopped = onStopped;
    onListening();

    await _speechToText.listen(
      onResult: (SpeechRecognitionResult result) {
        _lastRecognizedWords = result.recognizedWords;
        onResult(_lastRecognizedWords);
      },
      localeId: 'zh_CN',
      listenMode: ListenMode.confirmation,
      cancelOnError: true,
    );
  }

  Future<void> stopListening() async {
    await _speechToText.stop();
    _isListening = false;
    if (_onStopped != null) {
      _onStopped!();
      _onStopped = null;
    }
  }

  Future<void> cancelListening() async {
    await _speechToText.cancel();
    _isListening = false;
    if (_onStopped != null) {
      _onStopped!();
      _onStopped = null;
    }
  }

  Future<void> speak(String text) async {
    await _flutterTts.speak(text);
  }

  Future<void> stopSpeaking() async {
    await _flutterTts.stop();
  }

  void dispose() {
    _speechToText.cancel();
    _flutterTts.stop();
  }
}
